docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1163 @@
+1].AppState*; *chain* is defined in
+[**[TMBC-SEQ]**][TMBC-SEQ-link].
+
+#### **[FS-VAR-PEER-INV]**:
+It is always the case that the set *peerIDs* only contains nodes that
+have not yet misbehaved (by sending wrong data or timing out).
+
+#### **[FS-VAR-BLOCK-INV]**:
+For *startBlock.Height <= i < height - 1*, let *b(i)* be the block with
+height *i* in *blockstore*, it always holds that
+*VerifyCommit(b(i), b(i+1).Commit) = true*. This means that *height*
+can only be incremented if all blocks with lower height have been verified.
+
+
+## FastSync V2
+
+
+<!--
+> Basic data structures. Simplified, so that we can focus on the distributed
+algorithm here. If existing: link to Tendermint data structures, and mentioned
+if details were omitted. 
+---->
+
+### Outline
+
+<!--
+> Describe solution (in English), decomposition into functions, where communication to other components happens.
+---->
+
+The protocol is described in terms of functions that are triggered by
+(external) events. The implementation uses a scheduler and a
+de-multiplexer to deal with communicating with peers and to
+trigger the execution of these functions:
+
+- `QueryStatus()`: regularly (currently every 10sec; necessarily
+  interval greater than *2 Delta*) queries all peers from *peerIDs*
+  for their current height [TMBC-CORR-FULL]. It does so
+  by calling `Status(n)` remotely on all peers *n*.
+  
+- `CreateRequest`: regularly checks whether certain blocks have no
+  open request. If a block does not have an open request, it requests
+  one from a peer. It does so by calling `Block(n,h)` remotely on one
+  peer *n* for a missing height *h*.
+  
+> We have left the strategy how peers are selected unspecified, and
+> the currently existing different implementations of Fastsync differ
+> in this aspect. 
+
+The functions `Status` and `Block` are called by asynchronous
+RPC. When they return, the following functions are called:
+
+- `OnStatusResponse(addr Address, height int64)`: The full node with
+  address *addr* returns its current height. The function updates the height
+  information about *addr*, and may also increase *TargetHeight*.
+  
+- `OnBlockResponse(addr Address, b Block)`. The full node with
+  address *addr* returns a block. It is added to *blockstore*. Then
+  the auxiliary function `Execute` is called.
+
+- `Execute()`: Iterates over the *blockstore*.  Checks soundness of
+  the blocks, and
+  executes the transactions of a sound block and updates *state*. 
+
+> In addition to the functions above, the following two features are
+> implemented in Fastsync V2
+
+#### **[FS-V2-PEER-REMOVE]**:
+Periodically, *peerTimeStamp* and *peerRate* are analyzed. If a peer *p*
+has not provided a block recently (check of *peerTimeStamp[p]*) or it
+has not provided sufficiently many data (check of *peerRate[p]*), then
+*p* is removed from *peerIDs*.
+  
+#### **[FS-V2-TIMEOUT]**:
+
+*Fastsync V2* starts a timeout whenever a block is
```

Comment:
To be precise, implementation is slightly different. Node remember the last point in time when a block has been executed. Then there is an action that is triggered periodically and if no progress has been made for at least syncTimeout, i.e., (now - lastUpdate > syncTimeout) then fastSync terminates.

Reply:
We can describe it more precisely, but I would consider this an implementation detail of how it is checked whether a timeout occurred.